### ISSUE TITLE
A function that accepts a recursive ``type Thing = int | list[Thing]`` only accepts literals, not variables

### ISSUE THREAD

**Issue by Fuyukai** [microsoft/pyright#9798]
**Describe the bug**
A function that accepts a recursive ``type Thing = int | list[Thing]`` only accepts literals, not variables

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
type Thing = int | list[Thing]

def accepts_thing(thing: Thing):
    pass

accepts_thing(1)
accepts_thing([1, 2, 3])
lst = [1, 2, 3]
accepts_thing(lst)
```

If ``lst`` is changed to be ``lst: Thing`` or ``lst: list[Thing]`` then the example does not have a type error.

**VS Code extension or command-line**

VSCode

**Comment by erictraut** [comment]
Pyright is working correctly here and in conformance with the typing spec.

The type of variable `lst` in this case is inferred as `list[int]` by pyright's standard inference rules for lists. For details, refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=list-expressions).

The `list` class uses an invariant type parameter, which means that `list[int]` is not assignable to `list[Thing]`. This is necessary because `list` is mutable. For more details about mutable container types and variance, see [this documentation](https://microsoft.github.io/pyright/#/type-concepts?id=generic-types).

This works when you include the list expression as an argument to the `accepts_thing` call because pyright is able to use [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). This overrides the normal list inference rules for list expressions, allowing (in this case) for the expression to be inferred as `list[Thing]` rather than the normal `list[int]`.

You can employ bidirectional type inference in the case of the `lst` variable by providing an explicit type declaration, like this:

```python
lst: Thing = [1, 2, 3]
```
